GNB 함수 기능 적용 및 MobileNavigation 트랜지션 오류 수정

seongjin2427 opened the pull request:
# Pull Request

- [x] Fix bug

## Summary

- GNB 함수 기능 적용 및 `MobileNavigation` 트랜지션 오류 수정

## Describe your changes

- GNB 로그인, 회원가입 버튼을 위한 `LoginJoinList` 컴포넌트 추가
  - `GNB | Navi`에 따라 스타일 구분하여 MobileNavigation과 공용
    컴포넌트화 적용

- Resize 이벤트 시, `MobileNavigation`의 트랜지션이 밀려 따라 붙는 현상
  - css `opacity`를 0으로 줘서 토글 `false`일 때는 보이지 않게 수정

- `useToggle` 커스텀 훅 구현
  - 초기 값 받아서 `toggle` 기능 사용할 수 있도록 구현
  - `toggle`: 초기 값 `boolean`
  - `onToggle`: 매 클릭 시 `!boolean`
  - `setOn`: 클릭 시 `true`
  - `setOff`: 클릭 시 `false`

## Issue number and link

#48 

## ETC

현재 MobileNavigation 상태입니다.

![트리퐁 네비게이션](https://user-images.githubusercontent.com/59536977/185747079-be077865-496c-4f90-b9cd-46b3809dd778.gif)

seongjin2427 committed:
fix: GNB 함수 형태로 작동 수정

- MakeMenu util을 별도로 작성하여 로그아웃시에만 별도의 함수가
  실행되도록 구현

seongjin2427 committed:
fix: GNB 함수 기능 적용 및 MobileNavigation 트랜지션 오류 수정

- GNB 로그인, 회원가입 버튼을 위한 LoginJoinList 컴포넌트 추가
  - "GNB | Navi"에 따라 스타일 구분하여 MobileNavigation과 공용
    컴포넌트화 적용
- Resize 이벤트 시, MobileNavigation의 트랜지션이 밀려 따라 붙는 현상
  - Opacity를 주어 토글 false일 때는 보이지 않게 수정

sangrimlee commented on components/shared/LoginJoinList/LoginJoinList.tsx:
```diff
@@ -0,0 +1,33 @@
+import { LOGIN_MENUS } from '@/constants/menus';
+import { useRouter } from 'next/router';
+import React from 'react';
+import { useDispatch } from 'react-redux';
+import * as Styled from './LoginJoinList.styled';
```
나중에 import 순서도 어떤 규칙을 만들거나, prettier plugin을 통해서 정의해도 좋을 것 같아요!

seongjin2427 replied:
일단 정렬은 해두었습니다! ㅎㅎ

sangrimlee commented on constants/menus.ts:
```diff
@@ -1,23 +1,45 @@
+import { logoutUser } from '@/store/slice/userSlice';
+import { NextRouter } from 'next/router';
+import { Dispatch } from '@reduxjs/toolkit';
+
+interface ParamType {
```
조금 더 의미있는 이름을 주면 좋을 것 같아요.

seongjin2427 replied:
LoginJoinList 컴포넌트에서 들어오는 타입이다보니
LoginJoinParamsType이라고 수정해봤습니다

sangrimlee commented on layouts/MobileNaviation/MobileNaviation.styled.ts:
```diff
@@ -14,14 +14,15 @@ export const NavDiv = styled.div<NavDivProps>`
-  transform: none;
-  transition: transform ease-in-out 0.45s;
+  opacity: 0;
+  transform: translateX(100vw);
```
개인적으로는 viewport 단위를 사용하는 것에는 많은 어려움이 있다고 생각을 해요.

vw는 크게 문제가 되지 않는데 vh는 모바일 환경에서, 다양한 브라우저들이, 더 넓은 화면을 보여주기 위해서 계속해서 주소 표시창을 보여주었다가 없애주었다가가 반복이 되서 제대로 vh가 적용이 안된다는 문제점이 있습니다.

그래서, CSS2022에 올해 안에 적용될 CSS 기능중에 이 부분을 보완해주는 기능도 추가될 예정이에요. 하지만, 아직은 지원되지 않는 브라우저들이 많기 때문에, top, left, right, bottom으로 적용하는 편이 좋아보입니다.

예를 들어 지금과 같은 MobileNavgaion의 css를 다음과 같이 수정할 수 있을 것 같아요.
```css
position: fixed;
top: ${HEADER_HEIGHT};
left: 0;
right: 0;
bottom: 0;
transform: translateX(100%);
```

**Reference**
[CSS in 2022](https://wit.nts-corp.com/2022/02/24/6490) 의 New Viewport Units

seongjin2427 replied:
넵 반영했습니다!
CSS in 2022 사이트 공유 감사합니다!

jma1020 replied:
공유 감사합니다!

sangrimlee commented on layouts/GNB/GNB.tsx:
```diff
@@ -1,22 +1,31 @@
 
-function GNB({ isLogin }: GNBProps) {
+  const isLogin = useSelector(({ user }: AppState) => user.isLogIn);
   const [toggle, setToggle] = useState<boolean>(false);
```
자주 사용되는 toggle과 같은 상태를 useToggle이라는 hooks로 빼도 좋을 것 같아요!

seongjin2427 replied:
Header 컴포넌트이다보니 어느 페이지든 항상 존재하게 됩니다
그래서 다른 컴포넌트에도 토글이 필요할 경우, 
동시에 작동될 가능성이 있어보여 컴포넌트 내 State로 따로 두었습니다 ㅎㅎ

sangrimlee reviewed (approved):
고생하셨습니다!!

seongjin2427 committed:
refact: 피드백 반영

- LoginJoinList 컴포넌트 내 import 순서 정렬
- constants/menus.ts 내 ParamType interface명 변경 -> LoginJoinListParamType
- constants/menus.ts 내 LOGIN_MENUS onCilck 속성명 변경 -> excuteFn
- MobileNavigation 스타일 중 100vw 변경 -> 100% / left, right, bottom:
  0;

seongjin2427 committed:
feat: useToggle 커스텀 훅 구현 및 적용

- useToggle로 초기값 받아 [toggle, setToggle] 반환하여 사용할 수 있도록
구현

jma1020 reviewed (approved):
고생하셨습니다!! 😄

seongjin2427 committed:
feat: useToggle 수정

- useToggle 커스텀훅 객체형식으로 수정
 - toggle: 기본 boolean
 - onToggle: 클릭마다 !boolean
 - setOn: 클릭 시 true
 - setOff: 클릭 시 false
